app/controllers/uploads_controller.rb

```diff
@@ -53,9 +45,30 @@ def show
 
   private
 
+  def alert_messages(loaded_csv)
+    total_rows_count = loaded_csv.ids.length
+    failed_rows = loaded_csv.failed_instances
+    failed_rows_count = failed_rows.length
+    valid_rows = total_rows_count - failed_rows_count
+    validation_warnings = failed_rows.sort { |a, b| a.errors[:row].first.to_i <=> b.errors[:row].first.to_i }
+                                     .map(&:display_errors_with_row)
+    header_warnings = @upload.all_warnings
+
+    flash[:csv_success] = {
+      total_rows_count: total_rows_count.to_s,
+      valid_rows: valid_rows.to_s,
+      failed_rows_count: failed_rows_count.to_s
+    }.compact
+
+    flash[:warning] = {
+      'The following headers should be checked: ': (header_warnings unless header_warnings.empty?),
+      'The following rows should be checked: ': (validation_warnings unless validation_warnings.empty?)
+    }.compact
+  end
```

Comment:
I would love to see the calculations, i.e. 
- total_rows_count
- valid_rows
- failed_rows_count

be calculated in a Model somewhere. 

Also, I think that "'The following headers should be checked: '" & "'The following rows should be checked: '" is actually view code that belongs in the `_upload_messages` partial. This would require some tweaking of the data structure you're sharing between your controller and the view, but would be a bit cleaner.